src/supernova/utils.rs

```diff
@@ -23,51 +23,69 @@ use crate::{
 pub fn get_from_vec_alloc_relaxed_r1cs<G: Group, CS: ConstraintSystem<<G as Group>::Base>>(
   mut cs: CS,
   a: &[AllocatedRelaxedR1CSInstance<G>],
-  target_index: &AllocatedNum<G::Base>,
-) -> Result<(Vec<Boolean>, AllocatedRelaxedR1CSInstance<G>), SynthesisError> {
-  let mut a = a.iter().enumerate();
-  let mut selector: Vec<Boolean> = Vec::with_capacity(a.len());
-
-  let first_selected = alloc_num_equals_const(
-    cs.namespace(|| "check 0 equal bit".to_string()),
-    target_index,
-    G::Base::from(0),
-  )?;
-
-  selector.push(Boolean::Is(first_selected.clone()));
+  selector_vec: &[Boolean],
+) -> Result<AllocatedRelaxedR1CSInstance<G>, SynthesisError> {
+  assert_eq!(a.len(), selector_vec.len());
 
+  // Compare all instances in `a` to the first one
   let first = a
-    .next()
-    .ok_or_else(|| SynthesisError::IncompatibleLengthVector("empty vec length".to_string()))?
-    .1
-    .clone();
-
-  let initial_sum = Boolean::from(first_selected).lc(CS::one(), G::Base::ONE);
-
-  let (selected, selected_sum) = a.try_fold(
-    (first, initial_sum),
-    |(matched, mut selected_sum), (i, candidate)| {
-      let equal_bit = Boolean::Is(alloc_num_equals_const(
-        cs.namespace(|| format!("check {:?} equal bit", i)),
-        target_index,
-        G::Base::from(i as u64),
-      )?);
-
-      selector.push(equal_bit.clone());
-      selected_sum = selected_sum + &equal_bit.lc(CS::one(), G::Base::ONE);
+    .get(0)
+    .ok_or_else(|| SynthesisError::IncompatibleLengthVector("empty vec length".to_string()))?;
+
+  // Since `selector_vec` is correct, only one entry is 1.
+  // If selector_vec[0] is 1, then all `conditionally_select` will return `first`.
+  // Otherwise, the correct instance will be selected.
+  let selected = a
+    .iter()
+    .zip(selector_vec.iter())
+    .enumerate()
+    .skip(1)
+    .try_fold(first.clone(), |matched, (i, (candidate, equal_bit))| {
       let next_matched_allocated = conditionally_select_alloc_relaxed_r1cs(
         cs.namespace(|| format!("next_matched_allocated-{:?}", i)),
         candidate,
         &matched,
-        &equal_bit,
+        equal_bit,
       )?;
 
-      Ok::<(AllocatedRelaxedR1CSInstance<G>, LinearCombination<G::Base>), SynthesisError>((
-        next_matched_allocated,
-        selected_sum,
-      ))
-    },
-  )?;
+      Ok::<AllocatedRelaxedR1CSInstance<G>, SynthesisError>(next_matched_allocated)
+    })?;
+
+  Ok(selected)
+}
+
+/// Compute a selector vector `s` of size `num_indices`, such that
+/// `s[i] == 1` if i = `target_index` and 0 otherwise.
+/// Returns a SynthesisError if `target_index` is not in the range 0..num_indices.
+pub fn get_selector_vec_from_index<F: PrimeField, CS: ConstraintSystem<F>>(
+  mut cs: CS,
+  target_index: &AllocatedNum<F>,
+  num_indices: usize,
+) -> Result<Vec<Boolean>, SynthesisError> {
+  assert_ne!(num_indices, 0);
+
+  // Trivial case, return [1]
+  if num_indices == 1 {
+    return Ok(vec![Boolean::Constant(true)]);
+  }
```

Comment:
I'm not sure this should be a special case. We still need to ensure `target_index == 0` here, right?

Reply:
For the correctness of this function, I agree that we should then enforce `target_index == 0`.

However, as it is used in the circuit on the primary curve where there is only one running instance, we could entirely remove the allocation of `last_augmented_circuit_index` (make it optional) and use `[Boolean::Constant(true)]` as the selector without the additional constraint.

Reply:
Sure, I think they are orthogonal issues. In general, there is a tradeoff between complexity and constraint minimization. In cases where we're talking about saving one or two constraints at the cost of additional machinery and code paths, I would think hard before making that choice. The point I'm making is that if one were to choose the policy of *always* seeking minimal constraints, without regard for code complexity, then it's a safe bet that the result will be overwhelming. Let's take up the question of whether further optimization is warranted in a separate issue.

Reply:
That's fair, I removed it in the latest commit